tests/thread/thread.edl

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+enclave {
+    trusted {
+        public void cb_test_waiter_thread_impl();
+
+        public void cb_test_signal_thread_impl();
+
+        public void enc_test_mutex(
+            [in] size_t** count1,
+            [in] size_t** count2);
+
+        public void enc_wait(
+            size_t num_threads);
+
+        public void enc_signal();
+
+        public void enc_wait_for_exclusive_access();
+
+        public void enc_relinquish_exclusive_access();
+
+        public void enc_lock_and_unlock_mutexes(
+            [in, string] const char* mutex_ids);
+
+        public void enc_test_tcs_exhaustion(
+            [in] size_t** num_tcs_used,
+            [in] size_t** num_out_threads,
+            size_t tcs_req_count);
+
+        public void enc_reader_thread_impl(
+            [in] size_t** readers,
+            [in] size_t** writers,
+            [in] size_t** max_readers,
+            [in] bool** readers_and_writers);
+
+        public void enc_writer_thread_impl(
+            [in] size_t** readers,
```

Comment:
Can we clean up [in] size_t** ?
Instead what we want seems to be
[in,out] size_t readers[1]

Reply:
@EMumau Reading the test code a bit-more, my previous suggestion of  [in, out] size_t readers won't work since the variables are supposed to be updated by different threads. 

So one way I can think of is defining these variables in the enclave and then provide a query method for the host.

Or to have a an assert method that the host can call at the correct time and have the enclave perform all the assertions.

Reply:
No more double pointers.